File: exercises/clone_graph_133.java
import java.util.List;
import java.util.ArrayList;

// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) { return null; }
		var maximum_amount_of_unique_nodes = 101;
		var visited = new Node[maximum_amount_of_unique_nodes];
        cloneNode(node, visited);
        return visited[node.val];
    }

    private void cloneNode(Node node, Node[] visited) {
        Node clonedNode = new Node(node.val);
		visited[node.val] = clonedNode;
        for (Node neighbor : node.neighbors) {
            if (visited[neighbor.val] == null) {
                cloneNode(neighbor, visited);
            }
            visited[node.val].neighbors.add(visited[neighbor.val]);
        }
    }
}

File: exercises/clone_graph_133.ts
// IN THE OUTER FUNCTION
// null return null
// make heap array of 100 nodes
// call dfs with first node and visited array
// return the element from the array that corresponds to the index of the value of the first original node
//
// IN THE DFS
// clone the node
// add the cloned node to the array
// for each neighbor of node
// if its not in the visited array run dfs with current neighbor
// make the connection from copy node to copy node neighbor 


class _Node {
    val: number
    neighbors: _Node[]

    constructor(val?: number, neighbors?: _Node[]) {
        this.val = (val===undefined ? 0 : val)
        this.neighbors = (neighbors===undefined ? [] : neighbors)
    }
}

function cloneGraph(node: _Node | null): _Node | null {
	if (node == null) return null
	const max = 101
	const copied: (_Node|null)[] = new Array(max).fill(null)
	dfs(node, copied)
	return copied[node.val]
}

function dfs(node: _Node, copied: (_Node|null)[]) {
	const copy = new _Node(node.val)
	copied[node.val] = copy
	for (const neighbor of node.neighbors) {
		if (copied[neighbor.val] == null) {
			dfs(neighbor, copied)
		}
		copied[node.val]!.neighbors.push(copied[neighbor.val]!)
	}
}
